test/stdlib/Result.swift

```diff
@@ -0,0 +1,171 @@
+        return failure
+      case .success: 
+        expectUnreachable()
+        return nil
+    }
+  }()
+
+  expectEqual(string1, string)
+  expectEqual(error, .err)
+}
+
+ResultTests.test("Throwing Initialization and Unwrapping") {
+  func notThrowing() throws -> String {
+    return string
+  }
+
+  func throwing() throws -> String {
+    throw Err.err
+  }
+    
+  let result1 = Result { try throwing() }
+  let result2 = Result { try notThrowing() }
+  
+  expectEqual(result1.failure as? Err, Err.err)
+  expectEqual(result2.success, string)
+    
+  do {
+    _ = try result1.get()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    expectUnreachable()
+  }
+    
+  do {
+    let unwrapped = try result2.get()
+    expectEqual(unwrapped, string)
+  } catch {
+    expectUnreachable()
+  }
+    
+  // Test unwrapping strongly typed error.
+  let result3 = Result<String, Err>.failure(Err.err)
+  do {
+    _ = try result3.get()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    expectUnreachable()
+  }
+}
+
+ResultTests.test("Functional Transforms") {
+  func transformDouble(_ int: Int) -> Int {
+    return 2 * int
+  }
+  
+  func transformTriple(_ int: Int) -> Int {
+    return 3 * int
+  }
+  
+  func transformError(_ err: Err) -> Err {
+    if err == .err {
+      return .derr
+    } else {
+      return .err
+    }
+  }
+
+  func resultValueTransform(_ int: Int) -> Result<Int, Err> {
+    return .success(transformDouble(int))
+  }
+  
+  func resultErrorTransform(_ err: Err) -> Result<Int, Err> {
+    return .failure(transformError(err))
+  }
+    
+  let result1: Result<Int, Err> = .success(1)
+  let newResult1 = result1.map(transformDouble)
+    
+  expectEqual(newResult1, .success(2))
+    
+  let result2: Result<Int, Err> = .failure(.err)
+  let newResult2 = result2.mapError(transformError)
+    
+  expectEqual(newResult2, .failure(.derr))
+    
+  let result3: Result<Int, Err> = .success(1)
+  let newResult3 = result3.flatMap(resultValueTransform)
+    
+  expectEqual(newResult3, .success(2))
+    
+  let result4: Result<Int, Err> = .failure(.derr)
+  let newResult4 = result4.flatMapError(resultErrorTransform)
+    
+  expectEqual(newResult4, .failure(.err))
+}
+
+ResultTests.test("Equatable") {
+  let result1: Result<Int, Err> = .success(1)
+  let result2: Result<Int, Err> = .failure(.err)
+
+  expectEqual(result1, .success(1))
+  expectNotEqual(result1, .success(2))
+  expectNotEqual(result1, .failure(.err))
+  expectNotEqual(result1, .failure(.derr))
+
+  expectNotEqual(result2, .success(1))
+  expectNotEqual(result2, .success(2))
+  expectEqual(result2, .failure(.err))
+  expectNotEqual(result2, .failure(.derr))
+}
+
+ResultTests.test("Hashable") {
+  let result1: Result<Int, Err> = .success(1)
+  let result2: Result<Int, Err> = .success(2)
+  let result3: Result<Int, Err> = .failure(.err)
+  checkHashable([result1, result2, result3], equalityOracle: { $0 == $1 })
```

Comment:
It would also be useful to have a hashing check with potentially colliding Success/Failure hash encodings:

```suggestion
  checkHashable([result1, result2, result3], equalityOracle: { $0 == $1 })

  let confusables: [Result<Err, Err>] = [
    .success(.err)
    .success(.derr)
    .failure(.err)
    .failure(.derr)
  ]
  checkHashable(confusables, equalityOracle: { $0 == $1 })
```

`checkHashable` verifies that all four of these cases have different hash encodings. (The original implementation of `hash(into:)` had an issue with this, so it's not a theoretical concern.)

Reply:
Added.